File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/RandomWrapperMersenneTwisterRandomTest.java
package io.github.pr0methean.betterrandom.prng;

import com.google.common.collect.ImmutableList;
import io.github.pr0methean.betterrandom.seed.SeedException;
import io.github.pr0methean.betterrandom.seed.SeedGenerator;
import java.util.Random;
import org.testng.annotations.Test;

@Test(testName = "RandomWrapper:MersenneTwisterRandom")
public class RandomWrapperMersenneTwisterRandomTest extends MersenneTwisterRandomTest {

  private final NamedFunction<Random, Double> setWrapped;

  public RandomWrapperMersenneTwisterRandomTest() {
    SeedGenerator seedGenerator = getTestSeedGenerator();
    setWrapped = new NamedFunction<>(random -> {
      ((RandomWrapper) random).setWrapped(new MersenneTwisterRandom(seedGenerator));
      return 0.0;
    }, "setWrapped");
  }

  @Override public void testThreadSafety() {
    super.testThreadSafety();
    testThreadSafetyVsCrashesOnly(30,
        ImmutableList.of(NEXT_LONG, NEXT_INT, NEXT_DOUBLE, NEXT_GAUSSIAN, setWrapped));
  }

  @Override protected Class<? extends BaseRandom> getClassUnderTest() {
    return RandomWrapper.class;
  }

  @Override protected RandomWrapper createRng() throws SeedException {
    return new RandomWrapper(new MersenneTwisterRandom(getTestSeedGenerator()));
  }

  @Override protected RandomWrapper createRng(final byte[] seed) throws SeedException {
    return new RandomWrapper(new MersenneTwisterRandom(seed));
  }
}

File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/RandomWrapperMersenneTwisterRandomTest.kt
package io.github.pr0methean.betterrandom.prng

import com.google.common.collect.ImmutableList
import io.github.pr0methean.betterrandom.seed.SeedException
import io.github.pr0methean.betterrandom.seed.SeedGenerator
import java.lang.reflect.InvocationTargetException
import java.util.Random
import org.testng.annotations.Test

@Test(testName = "RandomWrapper:MersenneTwisterRandom")
class RandomWrapperMersenneTwisterRandomTest : MersenneTwisterRandomTest() {

    private val setWrapped: BaseRandomTest.NamedFunction<Random, Double>

    protected override val classUnderTest: Class<out BaseRandom>
        get() = RandomWrapper::class.java

    init {
        val seedGenerator = testSeedGenerator
        setWrapped = BaseRandomTest.NamedFunction({ random ->
            (random as RandomWrapper).wrapped = MersenneTwisterRandom(seedGenerator)
            0.0
        }, "setWrapped")
    }

    override fun testThreadSafety() {
        super.testThreadSafety()
        testThreadSafetyVsCrashesOnly(30,
                ImmutableList.of(BaseRandomTest.NEXT_LONG, BaseRandomTest.NEXT_INT, BaseRandomTest.NEXT_DOUBLE, BaseRandomTest.NEXT_GAUSSIAN, setWrapped))
    }

    @Test(enabled = false)
    @Throws(SeedException::class, IllegalAccessException::class, InstantiationException::class, InvocationTargetException::class)
    override fun testAllPublicConstructors() {
        // No-op: redundant to super insofar as it works.
    }

    @Throws(SeedException::class)
    override fun createRng(): RandomWrapper {
        return RandomWrapper(MersenneTwisterRandom(testSeedGenerator))
    }

    @Throws(SeedException::class)
    override fun createRng(seed: ByteArray?): RandomWrapper {
        return RandomWrapper(MersenneTwisterRandom(seed))
    }
}
